contracts/libraries/ERC20Splitter.sol

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+contract ERC20Splitter {
+  // Mapping to track balances of token by address and recipient
+  mapping(address => mapping(address => uint256)) private _balances;
+
+  // Mapping to track which tokens a user has received (for auto-withdrawals)
+  mapping(address => address[]) private _userTokens;
+  mapping(address => mapping(address => bool)) private _hasToken;
+
+  event Deposit(address indexed depositor, address[] tokenAddresses, uint256[] amounts, uint256[][] shares, address[][] recipients);
+  event Withdraw(address indexed recipient, address tokenAddress, uint256 amount);
+
+  /// @notice Deposits ERC20 or native tokens and splits between recipients based on shares.
+  /// @param tokenAddresses Array of token addresses (use address(0) for native tokens).
+  /// @param amounts Array of amounts for each token.
+  /// @param shares Array of share percentages (out of 10000) for each recipient.
+  /// @param recipients Array of recipients for each token.
+  function deposit(
+    address[] calldata tokenAddresses,
+    uint256[] calldata amounts,
+    uint256[][] calldata shares,
+    address[][] calldata recipients
+  ) external payable {
+    require(tokenAddresses.length == amounts.length, "Invalid input lengths");
+    require(tokenAddresses.length == shares.length && tokenAddresses.length == recipients.length, "Mismatched input sizes");
+
+    for (uint256 i = 0; i < tokenAddresses.length; i++) {
+      _splitTokens(tokenAddresses[i], amounts[i], shares[i], recipients[i]);
+    }
+
+    emit Deposit(msg.sender, tokenAddresses, amounts, shares, recipients);
+  }
+
+  /// @notice Withdraw all tokens that the caller is entitled to.
+  /// Tokens are automatically determined based on previous deposits.
+  function withdraw() external {
+    address[] storage userTokens = _userTokens[msg.sender];
+
+    require(userTokens.length > 0, "No tokens to withdraw");
+
+    for (uint256 i = 0; i < userTokens.length; i++) {
+      address tokenAddress = userTokens[i];
+      uint256 amount = _balances[tokenAddress][msg.sender];
+
+      if (amount > 0) {
+        _balances[tokenAddress][msg.sender] = 0;
+
+        if (tokenAddress == address(0)) {
+          payable(msg.sender).transfer(amount);
+        } else {
+          require(IERC20(tokenAddress).transfer(msg.sender, amount), "Transfer failed");
+        }
+
+        emit Withdraw(msg.sender, tokenAddress, amount);
+      }
+    }
+    _clearUserTokens(msg.sender);
```

Comment:
You are looping twice here, when you could have done it on the previous loop. Clean the tokens on the previous loop as well